syntax = "proto3";

package beam_grpc;

service BeamService {
  rpc query_algorithm(method_request) returns (pickled_response) {}
  rpc call_function(func_request) returns (pickled_response) {}
  rpc get_info(info_request) returns (info_response) {}
  rpc set_variable(set_variable_request) returns (set_variable_response) {}
  rpc get_variable(get_variable_request) returns (get_variable_response) {}
}

message func_request {
  bytes args = 1;  // Serialized arguments
  bytes kwargs = 2; // Serialized keyword arguments
}

message method_request {
  string method_name = 1;
  bytes args = 2;  // Serialized arguments
  bytes kwargs = 3; // Serialized keyword arguments
}

message pickled_response {
  bytes result = 1;  // Serialized result
}

message info_request {}

message info_response {
  string info_json = 1;
}

message set_variable_request {
  string name = 1;
  bytes value = 2;
}

message set_variable_response {
  bool success = 1;
}

message get_variable_request {
  string name = 1;
}

message get_variable_response {
  bytes value = 1;
}
